### ISSUE TITLE
BigQuery.Storage BigQueryReadClient ReadRows throws exception: Request is missing 'x-goog-request-params' header.

### ISSUE THREAD

**Issue by foresttian** [googleapis/google-cloud-dotnet#6310]
#### Environment details

  - OS: Windows
  - .NET version: 5.0
  - Package name and version: Google.Cloud.BigQuery.Storage.V1 2.1.0

#### Steps to reproduce

  1. Setup GOOGLE_APPLICATION_CREDENTIALS
  2.        
```
var readClient = BigQueryReadClient.Create();
var response = readClient.ReadRows("projects/atus-a4-prod-248221/locations/us/sessions/CAISDDdzb1habUVoaVdhVRoCanIaAmlyGgJvcxoCanEaAm93GgJqcxoCaXcaAm5hGgJqYxoCamQaAm9qGgJqZg/streams/GgJqchoCaXIaAm9zGgJqcRoCb3caAmpzGgJpdxoCbmEaAmpjGgJqZBoCb2oaAmpmKAI", 0L);
AsyncResponseStream<ReadRowsResponse> stream = response.GetResponseStream();
while (await stream.MoveNextAsync())
{
    ReadRowsResponse responseItem = stream.Current;
    // Do something with streamed response
}
```

  3. Error following will be found `{"Status(StatusCode=\"InvalidArgument\", Detail=\"Request is missing 'x-goog-request-params' header.\", DebugException=\"Grpc.Core.Internal.CoreErrorDetailException: {\"created\":\"@1617306048.491000000\",\"description\":\"Error received from peer ipv4:172.217.164.234:443\",\"file\":\"T:\\src\\github\\grpc\\workspace_csharp_ext_windows_x64\\src\\core\\lib\\surface\\call.cc\",\"file_line\":1062,\"grpc_message\":\"Request is missing 'x-goog-request-params' header.\",\"grpc_status\":3}\")"}`

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

**Comment by jskeet** [comment]
I'll have a look at this when I'm back at work on Tuesday - thanks for the report.

**Comment by foresttian** [comment]
Thanks!

**Comment by jskeet** [comment]
Okay, I've reproduced the bug. Normally the generated code includes an `x-goog-request-params` header, but we're not doing that for streaming calls such as `ReadRows`. I'm not yet sure whether we *should* be doing that or whether the BigQuery Storage API shouldn't be relying on it... however, there's a workaround until we've got that figured out.

You can create a `CallSettings` that specifies the header yourself, like this:

```csharp
var callSettings = CallSettings.FromHeader("x-goog-request-params",
    "read_stream=" + Uri.EscapeDataString(streamName.ToString()));
```
(Here `streamName` is a `ReadStreamName` - but you could just use a string and optionally remove the `ToString()` call.)

Pass the `CallSettings` into the `ReadRows` call, and it should work fine.

Here's a complete example which works for me:

```csharp
using Google.Api.Gax.Grpc;
using Google.Api.Gax.ResourceNames;
using Google.Cloud.BigQuery.Storage.V1;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace Issue6310
{
    class Program
    {
        static async Task Main()
        {
            var client = BigQueryReadClient.Create();

            // Adjust these for the table you want to read from
            string projectId = "...";
            string datasetId = "...";
            string tableId = "...";

            var readSession = new ReadSession
            {
                 TableAsTableName = new TableName(projectId, datasetId, tableId),
                 DataFormat = DataFormat.Avro
            };
            readSession = client.CreateReadSession(new ProjectName(projectId), readSession, 1);

            var streamName = readSession.Streams.First().ReadStreamName;
            Console.WriteLine($"Reading from stream {streamName}");

            var callSettings = CallSettings.FromHeader("x-goog-request-params",
                "read_stream=" + Uri.EscapeDataString(streamName.ToString()));
            var response = client.ReadRows(streamName, 0L, callSettings);
            AsyncResponseStream<ReadRowsResponse> stream = response.GetResponseStream();
            await foreach (var item in stream)
            {
                Console.WriteLine(item.RowCount);
            }
        }
    }
}
```

**Comment by foresttian** [comment]
Hi, I tried the callsettings approach. It works great. Thanks a lot for so quick response!

**Comment by jskeet** [comment]
This is fixed in 2.2.0 - you shouldn't need the CallSettings any more.

**Comment by foresttian** [comment]
Thanks for letting me know! I will have a try.
